Content.Server/Atmos/EntitySystems/GenericGasReactionSystem.cs

```diff
@@ -0,0 +1,115 @@
+using Content.Server.Atmos.Reactions;
+using Content.Shared.Atmos.Reactions;
+using Content.Shared.Atmos;
+using JetBrains.Annotations;
+using System.Collections;
+using System.Linq;
+
+namespace Content.Server.Atmos.EntitySystems;
+
+public sealed class GenericGasReactionSystem : EntitySystem
+{
+    [Dependency] private readonly AtmosphereSystem _atmosphere = default!;
+
+    /// <summary>
+    ///     Return a reaction rate (in units reactants per second) for a given reaction. Based on the
+    ///     Arrhenius equation (https://en.wikipedia.org/wiki/Arrhenius_equation).
+    ///
+    ///     This means that most reactions scale exponentially above the MinimumTemperatureRequirement.
+    /// </summary>
+    private float ReactionRate(GasReactionPrototype reaction, GasMixture mix, float dE)
+    {
+        float temp = mix.Temperature;
+
+        // Gas reactions have a MinimumEnergyRequirement which is in spirit activiation energy (Ea),
+        // but no reactions define it. So we have to calculate one to use. One way is to assume that
+        // Ea = 10*R*MinimumTemperatureRequirement such that Ea >> RT.
+        float TScaleFactor = 10;
+        float Ea = TScaleFactor*Atmospherics.R*reaction.MinimumTemperatureRequirement + dE;
+
+        // To compute initial rate coefficient A, assume that at temp = min temp we return 1/10.
+        float RateScaleFactor = 10; // not necessarily the same as TScaleFactor! Don't get confused!
+        float A = MathF.Exp(TScaleFactor) / RateScaleFactor;
+
+        // Prevent divide by zero
+        if (temp < Atmospherics.TCMB)
+            return 0;
+
+        return A*MathF.Exp(-Ea/(Atmospherics.R*temp));
+    }
+
+    /// <summary>
+    ///     Run all of the reactions given on the given gas mixture located in the given container.
+    /// </summary>
+    public ReactionResult ReactAll(IEnumerable<GasReactionPrototype> reactions, GasMixture mix, IGasMixtureHolder? holder)
+    {
+        float nTotal = mix.TotalMoles;
+        // Guard against very small amounts of gas in mixture
+        if (nTotal < Atmospherics.GasMinMoles)
```

Comment:
This should be checked for each of the reaction ingredients individually at some point. For example, you don't want to do the N2O Decomposition calculation in pure oxygen. Also, GasMinMoles = 0.00000005f is a really low threshold and you can probably keep the old ones of 0.01 moles per gas. If you have a big room with a tiny trace amount of a gas, you don't want to do the reaction on every single tile. This should ensure better numerical stability as well.

Reply:
This suggests to me that `GasMinMoles` should perhaps be raised in a separate PR. Or perhaps we should create a separate global set to 0.01 and make this a global minimum reaction amount for all reactions? It doesn't seem to make sense to me to allow different minimums for different reactions.